File: jdroid-android-core/src/debug/java/com/jdroid/android/debug/DebugContext.java
package com.jdroid.android.debug;

import android.app.Activity;
import android.support.v4.util.Pair;

import com.jdroid.android.activity.ActivityLauncher;
import com.jdroid.android.debug.appenders.ExceptionHandlingDebugPrefsAppender;
import com.jdroid.android.debug.appenders.HttpCacheDebugPrefsAppender;
import com.jdroid.android.debug.appenders.HttpMocksDebugPrefsAppender;
import com.jdroid.android.debug.appenders.InfoDebugPrefsAppender;
import com.jdroid.android.debug.appenders.NavDrawerDebugPrefsAppender;
import com.jdroid.android.debug.appenders.NotificationsDebugPrefsAppender;
import com.jdroid.android.debug.appenders.ServersDebugPrefsAppender;
import com.jdroid.android.debug.appenders.UriMapperPrefsAppender;
import com.jdroid.android.debug.appenders.UsageStatsDebugPrefsAppender;
import com.jdroid.java.collections.Lists;
import com.jdroid.java.collections.Maps;
import com.jdroid.java.http.Server;
import com.jdroid.java.remoteconfig.RemoteConfigParameter;

import java.util.List;
import java.util.Map;

public class DebugContext {

	private List<Pair<String, Object>> customDebugInfoProperties = Lists.newArrayList();

	public void launchActivityDebugSettingsActivity(Activity activity) {
		ActivityLauncher.startActivity(activity, DebugSettingsActivity.class);
	}

	public List<Pair<String, Object>> getCustomDebugInfoProperties() {
		return customDebugInfoProperties;
	}

	public void addCustomDebugInfoProperty(Pair<String, Object> pair) {
		customDebugInfoProperties.add(pair);
	}

	public ServersDebugPrefsAppender createServersDebugPrefsAppender() {
		return new ServersDebugPrefsAppender(getServersMap());
	}

	public Map<Class<? extends Server>, List<? extends Server>> getServersMap() {
		return Maps.newHashMap();
	}

	public ExceptionHandlingDebugPrefsAppender createExceptionHandlingDebugPrefsAppender() {
		return new ExceptionHandlingDebugPrefsAppender();
	}

	public HttpCacheDebugPrefsAppender createHttpCacheDebugPrefsAppender() {
		return new HttpCacheDebugPrefsAppender();
	}

	public NavDrawerDebugPrefsAppender createNavDrawerDebugPrefsAppender() {
		return new NavDrawerDebugPrefsAppender();
	}

	public HttpMocksDebugPrefsAppender createHttpMocksDebugPrefsAppender() {
		return new HttpMocksDebugPrefsAppender();
	}

	public InfoDebugPrefsAppender createInfoDebugPrefsAppender() {
		return new InfoDebugPrefsAppender();
	}

	public UsageStatsDebugPrefsAppender createUsageStatsDebugPrefsAppender() {
		return new UsageStatsDebugPrefsAppender();
	}

	public UriMapperPrefsAppender createUriMapperPrefsAppender() {
		return new UriMapperPrefsAppender();
	}

	public NotificationsDebugPrefsAppender createNotificationsDebugPrefsAppender() {
		return new NotificationsDebugPrefsAppender();
	}

	public List<PreferencesAppender> getCustomPreferencesAppenders() {
		return Lists.newArrayList();
	}

	public List<String> getUrlsToTest() {
		return Lists.newArrayList();
	}

	public List<RemoteConfigParameter> getRemoteConfigParameters() {
		return Lists.newArrayList();
	}
}

File: jdroid-android-core/src/debug/java/com/jdroid/android/debug/DebugContext.kt
package com.jdroid.android.debug

import android.app.Activity
import androidx.core.util.Pair

import com.jdroid.android.activity.ActivityLauncher
import com.jdroid.android.debug.appenders.ExceptionHandlingDebugPrefsAppender
import com.jdroid.android.debug.appenders.HttpCacheDebugPrefsAppender
import com.jdroid.android.debug.appenders.HttpMocksDebugPrefsAppender
import com.jdroid.android.debug.appenders.InfoDebugPrefsAppender
import com.jdroid.android.debug.appenders.NavDrawerDebugPrefsAppender
import com.jdroid.android.debug.appenders.NotificationsDebugPrefsAppender
import com.jdroid.android.debug.appenders.ServersDebugPrefsAppender
import com.jdroid.android.debug.appenders.UriMapperPrefsAppender
import com.jdroid.android.debug.appenders.UsageStatsDebugPrefsAppender
import com.jdroid.java.collections.Lists
import com.jdroid.java.collections.Maps
import com.jdroid.java.http.Server
import com.jdroid.java.remoteconfig.RemoteConfigParameter

open class DebugContext {

    private val customDebugInfoProperties = Lists.newArrayList<Pair<String, Any>>()

    fun launchActivityDebugSettingsActivity(activity: Activity) {
        ActivityLauncher.startActivity(activity, DebugSettingsActivity::class.java)
    }

    fun getCustomDebugInfoProperties(): List<Pair<String, Any>> {
        return customDebugInfoProperties
    }

    fun addCustomDebugInfoProperty(pair: Pair<String, Any>) {
        customDebugInfoProperties.add(pair)
    }

    open fun createServersDebugPrefsAppender(): ServersDebugPrefsAppender {
        return ServersDebugPrefsAppender(getServersMap())
    }

    open fun getServersMap(): Map<Class<out Server>, List<Server>> {
        return Maps.newHashMap()
    }

    open fun createExceptionHandlingDebugPrefsAppender(): ExceptionHandlingDebugPrefsAppender {
        return ExceptionHandlingDebugPrefsAppender()
    }

    open fun createHttpCacheDebugPrefsAppender(): HttpCacheDebugPrefsAppender {
        return HttpCacheDebugPrefsAppender()
    }

    open fun createNavDrawerDebugPrefsAppender(): NavDrawerDebugPrefsAppender {
        return NavDrawerDebugPrefsAppender()
    }

    open fun createHttpMocksDebugPrefsAppender(): HttpMocksDebugPrefsAppender {
        return HttpMocksDebugPrefsAppender()
    }

    open fun createInfoDebugPrefsAppender(): InfoDebugPrefsAppender {
        return InfoDebugPrefsAppender()
    }

    open fun createUsageStatsDebugPrefsAppender(): UsageStatsDebugPrefsAppender {
        return UsageStatsDebugPrefsAppender()
    }

    open fun createUriMapperPrefsAppender(): UriMapperPrefsAppender {
        return UriMapperPrefsAppender()
    }

    open fun createNotificationsDebugPrefsAppender(): NotificationsDebugPrefsAppender {
        return NotificationsDebugPrefsAppender()
    }

    open fun getCustomPreferencesAppenders(): List<PreferencesAppender> {
        return Lists.newArrayList()
    }

    open fun getUrlsToTest(): List<String> {
        return Lists.newArrayList()
    }

    open fun getRemoteConfigParameters(): List<RemoteConfigParameter> {
        return Lists.newArrayList()
    }
}
